binding.cc

```diff
@@ -600,8 +621,66 @@ struct Database {
 
 private:
   uint32_t priorityWork_;
+  std::string location_;
+
+  // for separation of concerns the threadsafe functionality was kept at the global
+  // level and made a friend so it is explict where the threadsafe boundary exists
+  friend leveldb::Status threadsafe_open(const leveldb::Options &options,
+                                         bool multithreading,
+                                         Database &db_instance);
+  friend leveldb::Status threadsafe_close(Database &db_instance);
 };
 
+
+leveldb::Status threadsafe_open(const leveldb::Options &options,
+                                bool multithreading,
+                                Database &db_instance) {
+  std::unique_lock<std::mutex> lock(handles_mutex);
+
+  auto it = db_handles.find(db_instance.location_);
+  if (it == db_handles.end()) {
+    // Database not opened yet for this location
+    LevelDbHandle handle = {nullptr, 0, multithreading};
+    leveldb::Status status = leveldb::DB::Open(options, db_instance.location_, &handle.db);
+    if (!status.ok()) {
+      return status;
+    }
+    handle.open_handle_count++;
+    db_instance.db_ = handle.db;
+    db_handles[db_instance.location_] = handle;
+    return leveldb::Status::OK();
+  }
+
+  if (!(it->second.multithreading && multithreading)) {
+    // Database already opened for this location that disallows multithreading
+    return leveldb::Status::InvalidArgument("Database already opened. Must set multithreading flag to true for all instances");
```

Comment:
Error codes are preferred over messages (on the JS side):

> Errors [..] will have a `code` property that is an uppercase string. Error codes will not change between major versions, but error messages will. Messages may also differ between implementations; they are free and encouraged to tune messages.
> — [`abstract-level`](https://github.com/Level/abstract-level#errors)

But I think we can generally simplify the logic here, and fully bypass the handles by default.

What do you think about:

```js
leveldb::Status threadsafe_open(const leveldb::Options &options,
                                bool multithreading,
                                Database &db_instance) {
  // Bypass lock and handles if multithreading is disabled
  if (!multithreading) {
    return leveldb::DB::Open(options, db_instance.location_, &db_instance.db_);
  }

  std::unique_lock<std::mutex> lock(handles_mutex);

  auto it = db_handles.find(db_instance.location_);

  if (it == db_handles.end()) {
    // Database not opened yet for this location, unless it was with
    // multithreading disabled, in which case we're expected to fail here.
    LevelDbHandle handle = {nullptr, 0};
    leveldb::Status status = leveldb::DB::Open(options, db_instance.location_, &handle.db);

    if (status.ok()) {
      handle.open_handle_count++;
      db_instance.db_ = handle.db;
      db_handles[db_instance.location_] = handle;
    }
    
    return status;
  }

  ++(it->second.open_handle_count);
  db_instance.db_ = it->second.db;

  return leveldb::Status::OK();
}

leveldb::Status threadsafe_close(Database &db_instance) {
  std::unique_lock<std::mutex> lock(handles_mutex);
  
  auto it = db_handles.find(db_instance.location_);

  if (it == db_handles.end()) {
    // Was not opened with multithreading enabled
    delete db_instance.db_;
  } else if (--(it->second.open_handle_count) == 0) {
    delete it->second.db;
    db_handles.erase(it);
  }

  // ensure db_ pointer is nullified in Database instance
  db_instance.db_ = NULL;

  return leveldb::Status::OK();
}
```

(I coded this on GitHub so may have mistakes)

Reply:
Sure.  If that is a better pattern lets go that direction.  Updated